### PR TITLE
feat: consider multiple root paths for liquidjs engine #294

### PR DESCRIPTION
**Pull request by devjayantmalik**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test/test-liquid.js [hunk 1]
```diff
@@ -481,3 +481,44 @@ test('fastify.view with liquid engine template that does not exist errors correc
...
+  })
+
+  fastify.register(require('../index'), {
+    engine: {
+      liquid: engine
+    }
+  })
+
+  fastify.get('/', (req, reply) => {
+    reply.view('./templates/index.liquid', data)
```
Comment (darkgl0w) [inline test/test-liquid.js:502]:
```suggestion
    reply.view('index.liquid', data)
```
If you provide only the template name you actually test your feature.
(As it is now your test pass in the main branch ^^).

**Review by darkgl0w — CHANGES_REQUESTED** [review]
Can you add a test case where templates are stored in two distinct directories please ?
Something like this :
```
./
├─ template_root_one/
│  ├─ hello.liquid
├─ template_root_two/
│  ├─ world.liquid
```

File: test/test-liquid.js (hunk 1 above)
Comment (devjayantmalik) [reply test/test-liquid.js]: Please feel free to see the updated code in next push

**Review by mcollina — APPROVED** [review]
lgtm

File: test/test-liquid.js (hunk 1 above)
Comment (darkgl0w) [reply test/test-liquid.js]: Thank you :)
I will take a look in a few minutes :)

File: index.js [hunk 2]
```diff
@@ -533,7 +534,17 @@ function fastifyView (fastify, opts, next) {
...
+    let pagepath = join(templatesDir, page)
+    // Don't execute it if root path is included, then serve it directly
+    const normalised = (root || []).map(loc => normalize(loc))
+    const isRootIncluded = normalised.find(loc => page.indexOf(loc) !== -1)
+    if (typeof root.length !== 'undefined' && !isRootIncluded) {
+      const paths = root.map((dir) => join(resolve(dir), page))
+      const found = paths.find((filepath) => existsSync(filepath))
+      if (found) pagepath = found
+    }
+    engine.renderFile(pagepath, data, opts)
```
Comment (darkgl0w) [inline index.js:547]:
```suggestion
    const { root } = engine.options

    // Don't execute it if root path is included, then serve it directly
    const isRootIncluded = (root || [])
      .map((path) => normalize(path))
      .find((path) => page.indexOf(path) !== -1)

    const pagePath = (typeof root.length !== 'undefined' && !isRootIncluded)
      ? root
          .map((dir) => join(resolve(dir), page))
          .find((filepath) => existsSync(filepath)) || join(templatesDir, page)
      : join(templatesDir, page)

    engine.renderFile(pagePath, data, opts)
```

**Review by darkgl0w — APPROVED** [review]
LGTM

Refactored a bit to make it more readable (that's a personal preference tough)

NB: there are two branches that are not covered by tests:
- when root const is falsy
- when the file path is not found

**Comment by devjayantmalik** [comment]
I have added another test case as suggested by @darkgl0w

File: index.js [hunk 3]
```diff
@@ -533,7 +534,20 @@ function fastifyView (fastify, opts, next) {
...
+
+    // Don't execute it if root path is included, then serve it directly
+    const isRootIncluded = root
+      .map((path) => normalize(path))
+      .find((path) => page.indexOf(path) !== -1)
+
+    const pagePath = (typeof root.length !== 'undefined' && !isRootIncluded)
+      ? root
+          .map((dir) => join(resolve(dir), page))
+          .find((filepath) => existsSync(filepath)) || join(templatesDir, page)
```
Comment (Eomm) [inline index.js:547]: The `existsSync` function is blocking the event loop: you can accept any other requests while the process is checking if the file exists or not.

we have two options:

- call an `async` function
- if we can't call an `async` function (not sure why), we should cache any previous check. For example, the `index` template will be called many times from the homepage site.. so it is a waste of resource to check every time

**Review by Eomm — CHANGES_REQUESTED** [review]

**Comment by devjayantmalik** [comment]
I think we should get rid of this functionality completely and care about it in future, because currently nodejs doesn't support async file checks or we should find some better alternative.

Moreover I think the module is quite unstable, I will work on this module in future, whenever I get time. Till then Work it out guys!

Here is how:

Also, there is a issue with `point-of-view` package, it doesn't consider liquidJS options root path, rather it works, only if you provide root path to POV options.

Here is what I mean:

```js
    const engine = new Liquid({
      root: "./templates",
    });
    fastifyServer.register(POV, {
      engine: { liquid: engine },
    });

// Doesn't work because root is still ./
res.view("hello.liquid");

// It works because root dir inside package is still ./ it is not overwritten by liquid root path.
res.view("./templates/hello.liquid");
```

Here is the workaround solution to above problem:

- set the root path while registering POV.
```js
 const engine = new Liquid();
    fastifyServer.register(POV, {
      engine: { liquid: engine },
     root:  "./templates"

    });

// It works because root is ./templates
res.view("hello.liquid");

// It doesn't work because root dir inside package is still ./templates and it becomes: ./templates/templates/hello.liquid
res.view("./templates/hello.liquid");
```

**Comment by mcollina** [comment]
> I think we should get rid of this functionality completely and care about it in future, 

What functionality? Should we close this PR?

> because currently nodejs doesn't support async file checks or we should find some better alternative.

I don't understand this.

**Comment by Eomm** [comment]
> because currently nodejs doesn't support async file checks or we should find some better alternative.

You can use [`fs.access()`](https://nodejs.org/api/fs.html#fspromisesaccesspath-mode) to check it.

I hope your statements do not come as a reaction to my PR review 😞

**Pull request closed without merging**